File: nwerc2025/bisectingbargain/submissions/accepted/jeroenodbDumb.py
import sys


def solve(n):
    num =    [500,200,100,50,20,10,5,2,1]
    atmost = [999,999,  1, 1, 4, 1,1,4,1]
    ans = []
    coins = []
    def check():
        nonlocal coins, n
        if n%2==1: return True
        dp = [0]*(n//2+1)
        dp[0]=1
        for c in coins:
            for j in range(n//2+1-c)[::-1]:
                if dp[j]: dp[j+c]=1
        return not dp[n//2]
    iters=0
    def rec(at,left):
        
        nonlocal coins,n,ans,iters
        iters+=1
        if at==len(num):
            if left==0:
                if check():
                    ans=list(i for i in coins)
                    return True
                else:
                    return False
            return False
        cur=len(coins)
        for take in range(atmost[at]+1):
            if take*num[at]>left:
                break
            if rec(at+1,left-take*num[at]): return True
            coins.append(num[at])
        while len(coins)>cur:
            coins.pop()
        return False
    if not rec(0,n):
        print(iters,file=sys.stderr)

        print("splittable")
    else:
        print(iters,file=sys.stderr)

        print(len(ans))
        print(*ans)
n = int(input())
solve(n)

File: nwerc2025/bisectingbargain/submissions/accepted/jeroenodbDumb.cpp
#pragma GCC optimize("O3")
#include "bits/stdc++.h"
using namespace std;

typedef vector<int> vi;
typedef vector<vi> vvi;
typedef pair<int,int> pi;
const int oo = 1e9;

void solve(int n) {
    vi num =    {500,200,100,50,20,10,5,2,1};
    vi atmost = {999,999,  1, 1, 4, 1,1,4,1};
    vi ans;
    typedef vector<bool> DP;
    vi coins;
    auto check = [&]() {
        if(n%2==1) return true;
        DP dp(n/2+1);
        dp[0]=1;
        for(int c : coins) {
            for(int j=n/2-c;j>=0;--j) {
                if(dp[j]) {
                    dp[j+c]=1;
                }
            }
        }
        return !dp[n/2];
    };
    auto rec = [&](auto&& self, int at,int left) {
        if(at==size(num)) {
            if(left==0) {
                if(check()) {
                    ans=coins;
                    return true;
                } else {
                    return false;
                }
            }
            return false;
        }
        int cur=size(coins);
        for(int take=0;take<=atmost[at] and take*num[at]<=left;++take) {
            if(self(self,at+1,left-take*num[at])) {
                return true;
            }
            coins.push_back(num[at]);

        }
        coins.resize(cur);
        return false;
    };

    if(!rec(rec,0,n)) {
        cout << "splittable\n";
    } else {
        cout << size(ans);
        for(auto x : ans) cout << ' ' << x;
        cout << '\n';
    }
}

int main() {
    int n; cin >> n;
    solve(n);
}